File: dotnet/src/DocGen/Clients/XRechnungClient.cs
using Dokmatiq.DocGen.Internal;
using Dokmatiq.DocGen.Models;

namespace Dokmatiq.DocGen.Clients;

/// <summary>
/// Client for XRechnung operations.
/// </summary>
public sealed class XRechnungClient
{
    private readonly Transport _transport;

    internal XRechnungClient(Transport transport) => _transport = transport;

    /// <summary>Generate XRechnung XML from invoice data.</summary>
    public string Generate(InvoiceData invoiceData, XRechnungFormat? format = null)
    {
        var body = new Dictionary<string, object?> { ["invoiceData"] = invoiceData };
        if (format is not null) body["format"] = format;
        var result = _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post,
            "/api/xrechnung/generate", body);
        return result.TryGetValue("xml", out var xml) ? xml?.ToString() ?? "" : "";
    }

    /// <summary>Parse XRechnung XML into structured data.</summary>
    public InvoiceData Parse(string xml)
        => _transport.RequestJson<InvoiceData>(HttpMethod.Post, "/api/xrechnung/parse", new { xml });

    /// <summary>Validate XRechnung XML.</summary>
    public Dictionary<string, object> Validate(string xml)
        => _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post, "/api/xrechnung/validate", new { xml });

    /// <summary>Transform between XRechnung formats.</summary>
    public string Transform(string xml, XRechnungFormat targetFormat)
    {
        var result = _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post,
            "/api/xrechnung/transform", new { xml, targetFormat });
        return result.TryGetValue("xml", out var x) ? x?.ToString() ?? "" : "";
    }

    /// <summary>Detect if XML is XRechnung.</summary>
    public Dictionary<string, object> Detect(string xml)
        => _transport.RequestJson<Dictionary<string, object>>(HttpMethod.Post, "/api/xrechnung/detect", new { xml });

    /// <summary>Extract invoice data using AI.</summary>
    public ExtractionResult ExtractAI(string xml)
        => _transport.RequestJson<ExtractionResult>(HttpMethod.Post, "/api/xrechnung/extract-ai", new { xml });
}

File: java/src/main/java/com/dokmatiq/docgen/client/XRechnungClient.java
package com.dokmatiq.docgen.client;

import com.dokmatiq.docgen.internal.Transport;
import com.dokmatiq.docgen.model.ExtractionResult;
import com.dokmatiq.docgen.model.InvoiceData;
import com.dokmatiq.docgen.model.XRechnungFormat;

import java.util.HashMap;
import java.util.Map;

/** Client for XRechnung operations. */
public class XRechnungClient {
    private final Transport transport;

    public XRechnungClient(Transport transport) {
        this.transport = transport;
    }

    /** Generate XRechnung XML from invoice data. */
    public String generate(InvoiceData invoiceData, XRechnungFormat format) {
        var body = new HashMap<String, Object>();
        body.put("invoiceData", invoiceData);
        if (format != null) body.put("format", format);
        var result = transport.requestJson("POST", "/api/xrechnung/generate", body, Map.class);
        return String.valueOf(result.get("xml"));
    }

    /** Generate XRechnung XML (default format). */
    public String generate(InvoiceData invoiceData) {
        return generate(invoiceData, null);
    }

    /** Parse XRechnung XML into structured data. */
    public InvoiceData parse(String xml) {
        return transport.requestJson("POST", "/api/xrechnung/parse",
                Map.of("xml", xml), InvoiceData.class);
    }

    /** Validate XRechnung XML. */
    public Map<?, ?> validate(String xml) {
        return transport.requestJson("POST", "/api/xrechnung/validate",
                Map.of("xml", xml), Map.class);
    }

    /** Transform between XRechnung formats. */
    public String transform(String xml, XRechnungFormat targetFormat) {
        var result = transport.requestJson("POST", "/api/xrechnung/transform",
                Map.of("xml", xml, "targetFormat", targetFormat), Map.class);
        return String.valueOf(result.get("xml"));
    }

    /** Detect if XML is XRechnung. */
    public Map<?, ?> detect(String xml) {
        return transport.requestJson("POST", "/api/xrechnung/detect",
                Map.of("xml", xml), Map.class);
    }

    /** Extract invoice data using AI. */
    public ExtractionResult extractAI(String xml) {
        return transport.requestJson("POST", "/api/xrechnung/extract-ai",
                Map.of("xml", xml), ExtractionResult.class);
    }
}
